convert2rhel/unit_tests/subscription_test.py

```diff
@@ -463,6 +347,497 @@ def __call__(self, pkg, dest, reposdir=None):
+        assert utils.let_user_choose_item.called == 0
+
+    def test_attach_subscription_multiple_subs_available(self, monkeypatch):
+        monkeypatch.setattr(subscription, "get_avail_subs", GetAvailSubsMocked())
+        monkeypatch.setattr(utils, "let_user_choose_item", LetUserChooseItemMocked())
+        monkeypatch.setattr(utils, "run_subprocess", RunSubprocessMocked())
+
+        assert subscription.attach_subscription() is True
+        assert utils.let_user_choose_item.called == 1
+
+    def test_attach_subscription_available_with_activation_key(self, monkeypatch, caplog):
+        monkeypatch.setattr(subscription, "get_avail_subs", GetAvailSubsMocked())
+        monkeypatch.setattr(utils, "let_user_choose_item", LetUserChooseItemMocked())
+        monkeypatch.setattr(utils, "run_subprocess", RunSubprocessMocked())
+        monkeypatch.setattr(toolopts.tool_opts, "activation_key", "dummy_activation_key")
+
+        assert subscription.attach_subscription() is True
+        assert len(caplog.records) == 1
+        assert caplog.records[0].levelname == "INFO"
+
+    def test_attach_subscription_none_available(self, monkeypatch):
+        monkeypatch.setattr(subscription, "get_avail_subs", GetNoAvailSubsMocked())
+
+        assert subscription.attach_subscription() is False
+
+
+class TestRegisterSystem(object):
+    def test_register_system_fail_non_interactive(self, tool_opts, monkeypatch, caplog):
+        """Check the critical severity is logged when the credentials are given on the cmdline but registration fails."""
+        monkeypatch.setattr(subscription, "MAX_NUM_OF_ATTEMPTS_TO_SUBSCRIBE", 1)
+        monkeypatch.setattr(subscription, "sleep", mock.Mock())
+
+        fake_spawn = mock.Mock()
+        fake_spawn.before.decode = mock.Mock(return_value="nope")
+        fake_spawn.exitstatus = 1
+        monkeypatch.setattr(utils, "PexpectSizedWindowSpawn", fake_spawn)
+
+        tool_opts.username = "user"
+        tool_opts.password = "pass"
+        tool_opts.credentials_thru_cli = True
+
+        with pytest.raises(SystemExit):
+            subscription.register_system()
+
+        assert caplog.records[-1].levelname == "CRITICAL"
+
+    def test_register_system_fail_interactive(self, tool_opts, monkeypatch, caplog):
+        """Check the function tries to register multiple times without critical log."""
+        tool_opts.credentials_thru_cli = False
+        monkeypatch.setattr(subscription, "sleep", mock.Mock())
+
+        fake_from_tool_opts = mock.Mock(
+            return_value=subscription.RegistrationCommand(username="invalid", password="invalid")
+        )
+        monkeypatch.setattr(subscription.RegistrationCommand, "from_tool_opts", fake_from_tool_opts)
+
+        class FakeProcess(mock.Mock):
+            called_count = 0
+
+            @property
+            def exitstatus(self):
+                self.called_count += 1
+                return self.called_count % 3
```

Comment:
Just a minor suggestion for this: 

This could be moved to the outside test method, like the others class mocks to follow the same “pattern” and then, the test method itself would look more clean. 

No need to apply this if you don't think it makes sense.

Reply:
I think in this case, it's specific to the function being tested (It returns the non-zero exit codes 1 and 2 the first two times it's called and then a success (0) exit code the third time).  It could be made more generic and then moved to the toplevel, though.  But if so, I think it should take these things into consideration:

* How many iterations before success?
* Spawn makes available exitstatus and signalstatus, should this allow setting both?
* Should it also set stdout output?
* Should the specific exitstatus per invocation be settable or is using ` count % success_attempt` good enough?

I think those are all good things to consider but maybe we should save moving it to a subsquent PR.

Reply:
Hmm, yeah, agreed! I think we can include a comment then somewhere saying that if there is any need to reuse this class in another test, then those points you mentioned should be considered as well. 

Maybe in the future there will be more tests regarding pexpect that we need to do, so this might come in hand.